c7n/tags.py

```diff
@@ -917,6 +919,125 @@ def process_resource_set(self, resource_set, tags):
+    all tags from the related resource should be copied to the original resource.
+
+    To raise an error when related resources cannot be found, use the
+    ``skip_missing`` option. By default, this is set to True.
+
+    :example:
+
+      .. code-block :: yaml
+
+        policies:
+          - name: copy-tags-from-ebs-volume-to-snapshot
+            resource: ebs-snapshot
+            actions:
+              - type: copy-related-resource-tag
+                resource: ebs
+                skip_missing: True
+                key: VolumeId
+                copy_tags:
+                  - *
+    """
+
+    schema = utils.type_schema(
+        'copy-related-resource-tag',
+        resource={'type': 'string'},
+        skip_missing={'type': 'boolean'},
+        key={'type': 'string'},
+        copy_tags={'type': 'array'},
+        required=['copy_tags', 'key', 'resource']
+    )
+
+    def get_permissions(self):
+        return self.manager.action_registry.get('tag').permissions
+
+    def validate(self):
+        related_resource = self.data['resource']
+        if related_resource not in all_resources.keys():
+            raise PolicyValidationError(
+                "Error: Invalid resource type selected: %s" % related_resource
+            )
+        # ideally should never raise here since we shouldn't be applying this
+        # action to a resource if it doesn't have a tag action implemented
+        if self.manager.action_registry.get('tag') is None:
+            raise PolicyValidationError(
+                "Error: Tag action missing on resource"
+            )
+        return self
+
+    def process(self, resources):
+        tag_keys = self.data['copy_tags']
+        related_key = self.data['key']
+        related_type = self.data['resource']
+
+        related_resource_tuple = list(
+            zip(jmespath.search('[].%s' % related_key, resources), resources))
+
+        related_ids = [r[0] for r in related_resource_tuple]
+        related_tag_map = self.get_resource_tag_map(related_type, related_ids)
+        missing_related_tags = list(set(related_ids).difference(related_tag_map.keys()))
+
+        if not self.data.get('skip_missing', True) and missing_related_tags:
+            raise PolicyExecutionError(
+                "Unable to find all %s resources associated with %s" % (related_type, related_ids))
+
+        for related, r in related_resource_tuple:
+            if related in missing_related_tags or not related_tag_map[related]:
+                self.log.warning(
+                    'Tags not found for related resource: %s, skipping %s' % (related, r))
+                continue
+            related_tags = self.format_tags(related_tag_map[related])
+            self.process_resource(r, related_tags, tag_keys)
+
+    def process_resource(self, r, related_tags, tag_keys):
+        add_tags = {}
+        if '*' in tag_keys:
+            add_tags = related_tags
+        else:
+            for t in tag_keys:
```

Comment:
this can reduce to a one liner `add_tags = {}`

also its really change_tags as it encompasses both updates and changes.